### PR TITLE
Add local time-stepping for CCE

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

Performs some restructuring of the storage of time-stepper related tags -- moving several from the global cache to the DataBox, and enables the use of step choosers for the CCE system.
I anticipate that most of the time-related tags have sufficiently small memory footprints that it doesn't make much of a difference that they are in the DataBox, but I'm happy to consider alternative methods of approaching the structuring of the systems -- my main design constraint was that CCE evolution and DG element array should be able to simultaneously access the time-stepping architecture with their own respective time steppers and step choosers.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Dependencies

- [x] PR #3304

### CODE REVIEW

File: tests/Unit/Time/Actions/Test_ChangeStepSize.cpp [hunk 1]
```diff
@@ -35,6 +36,30 @@
 
 // IWYU pragma: no_forward_declare ActionTesting::InitializeDataBox
 
+namespace detail {
```
Comment (wthrowe) [inline tests/Unit/Time/Actions/Test_ChangeStepSize.cpp]: You don't need to use `detail` namespaces in tests.  Anonymous should be fine.

**Review by wthrowe — CHANGES_REQUESTED** [review]
No complaints from me about moving the time steppers and stuff.

File: tests/Unit/Time/Actions/Test_ChangeStepSize.cpp (hunk 1 above)
Comment (moxcodes) [reply tests/Unit/Time/Actions/Test_ChangeStepSize.cpp]: When it was an in an anonymous namespace, I ended up with surprising linker complaints about undefined entities with internal linkage. If that doesn't sound familiar, I'll rerun to get the exact linker error.

Comment (wthrowe) [reply tests/Unit/Time/Actions/Test_ChangeStepSize.cpp]: That's weird.  I think it should work.

If you do end up needing a named namespace, use something unique like `Test_ChangeStepSize_detail`.

**Comment by wthrowe** [comment]
Squash and rebase.

**Comment by moxcodes** [comment]
Squashed and rebased. I didn't see the linker error when running again -- I must have been doing something else wrong somewhere.
Cheers!

**Review by wthrowe — DISMISSED** [review]

File: src/Time/Actions/ChangeStepSize.hpp [hunk 2]
```diff
@@ -40,9 +40,17 @@ bool change_step_size(
...
 
   const auto& next_time_id = db::get<Tags::Next<Tags::TimeStepId>>(*box);
-  const auto& history = db::get<Tags::HistoryEvolvedVariables<>>(*box);
-
-  if (not time_stepper.can_change_step_size(next_time_id, history)) {
+  using history_tags = ::Tags::get_all_history_tags<DbTags>;
+  bool can_change_step_size = true;
+  tmpl::for_each<history_tags>([&box, &can_change_step_size, &time_stepper,
+                                &next_time_id](auto tag_v) noexcept {
+    using tag = typename decltype(tag_v)::type;
```
Comment (nilsdeppe) [inline src/Time/Actions/ChangeStepSize.hpp:50]: what do you think of writing this as:
```cpp
if (not can_change_step_size) {
  return;
}
using tag = typename decltype(tag_v)::type;
const auto& history = db::get<tag>(*box);
can_change_step_size = time_stepper.can_change_step_size(next_time_id, history);
```
? I personally find this pattern easier to understand, but completely optional :)

File: src/Evolution/DiscontinuousGalerkin/Actions/ComputeTimeDerivative.hpp [hunk 3]
```diff
@@ -288,14 +288,9 @@ struct ComputeTimeDerivative {
   using inbox_tags =
```
Comment (nilsdeppe) [inline src/Evolution/DiscontinuousGalerkin/Actions/ComputeTimeDerivative.hpp:288]: If you end up changing anything else there's a type in the middle of the second paragraph: "the a given collection" -> "a given collection"

**Review by nilsdeppe — DISMISSED** [review]
LGTM! One very optional idea. If you like it and feel up to it, please implement, otherwise we can merge this as-is :)

**Comment by moxcodes** [comment]
Squashed in the last couple of suggestions, fixed a global cache tag that needed to be removed following a rebase to get CI to pass.
Cheers!

**Review by wthrowe — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**